libsignal-service/src/service_address.rs

```diff
@@ -33,31 +36,93 @@ impl ServiceAddress {
     pub fn pni(&self) -> libsignal_protocol::Pni {
         libsignal_protocol::Pni::from_uuid_bytes(self.uuid.into_bytes())
     }
+
+    pub fn to_service_id(&self) -> String {
+        match self.identity {
+            ServiceIdType::AccountIdentity => self.uuid.to_string(),
+            ServiceIdType::PhoneNumberIdentity => {
+                format!("PNI:{}", self.uuid)
+            },
+        }
+    }
 }
 
 impl From<Uuid> for ServiceAddress {
     fn from(uuid: Uuid) -> Self {
-        Self { uuid }
+        Self {
+            uuid,
+            identity: ServiceIdType::AccountIdentity,
+        }
+    }
+}
+
+impl TryFrom<&ProtocolAddress> for ServiceAddress {
+    type Error = ParseServiceAddressError;
+
+    fn try_from(addr: &ProtocolAddress) -> Result<Self, Self::Error> {
+        let value = addr.name();
+        if let Some(pni) = value.strip_prefix("PNI:") {
+            Ok(ServiceAddress {
+                uuid: Uuid::parse_str(pni)?,
+                identity: ServiceIdType::PhoneNumberIdentity,
+            })
+        } else {
+            match Uuid::parse_str(value) {
+                Ok(uuid) => Ok(ServiceAddress {
+                    uuid,
+                    identity: ServiceIdType::AccountIdentity,
+                }),
+                Err(e) => {
+                    tracing::error!("Unknown identity format: {}", value);
+                    Err(ParseServiceAddressError::InvalidUuid(e))
+                },
+            }
+        }
     }
 }
 
 impl TryFrom<&str> for ServiceAddress {
     type Error = ParseServiceAddressError;
 
     fn try_from(value: &str) -> Result<Self, Self::Error> {
-        Ok(ServiceAddress {
-            uuid: Uuid::parse_str(value)?,
-        })
+        if value.starts_with("PNI:") {
+            Ok(ServiceAddress {
+                uuid: Uuid::parse_str(value.strip_prefix("PNI:").unwrap())?,
+                identity: ServiceIdType::PhoneNumberIdentity,
+            })
+        } else {
+            Ok(ServiceAddress {
+                uuid: Uuid::parse_str(value)?,
+                identity: ServiceIdType::AccountIdentity,
+            })
+        }
+    }
+}
+
+impl TryFrom<&[u8]> for ServiceAddress {
+    type Error = ParseServiceAddressError;
+
+    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+        if let Some(pni) = value.strip_prefix(b"PNI:") {
+            Ok(ServiceAddress {
+                uuid: Uuid::from_slice(pni)?,
+                identity: ServiceIdType::PhoneNumberIdentity,
+            })
+        } else {
+            Ok(ServiceAddress {
+                uuid: Uuid::from_slice(value)?,
+                identity: ServiceIdType::AccountIdentity,
+            })
+        }
     }
 }
 
 impl TryFrom<Option<&str>> for ServiceAddress {
```

Comment:
We should probably just deprecate or remove those two...

Reply:
IIRC `try_from` was used in some places and it resulted in quite a nice and Rust-y code. I'll see what happens when I try to remove those.